docs/generate/directives.md

```diff
@@ -0,0 +1,281 @@
+
+We usually recommend changing the swagger before going the directive route, but if you'd like the original swagger to remain untouched (for example, you want to rename the generated model in the Python SDK, but not in other SDKs), directives are the route for you.
+
+>Stylistic note: we recommend annotating your directives in your config file with a header about what the directive is doing.
+
+## Structure and Terminology
+
+Directives consist of three parts:
+
+- **Location**: denoted by the field `from`, which document are we trying to transform. For swagger transformations, it's always `from: swagger-document`.
+
+- **Filter**: denoted by the field `where`, contains the criteria to select the object.
+  - An `operation` is filtered by
+    - its path in the swagger's [`paths`][paths] object AND
+    - it's HTTP verb
+  - A `parameter` is filtered by:
+    - its location in its `operation` (see above) OR
+    - its name in the swagger's `parameters` object, if it's defined as a common parameter there (see the Common Parameters for Various Paths section [here][parameters] for more information)
+  - A `model` can be filtered by:
+    - its name in the [components or definitions][components] section OR
+    - its location in its outer object, if defined within another object OR
+    - its location in an operation, if defined within an operation
+  - A `property` can be filtered by:
+    - its location within its parent object
+
+- **Transform**: denoted by the field `transform`, the actions we would like to be applied on the specified objects.
+
+## Directive Scenarios
+
+The following directives cover the most common tweaking scenarios for generation.
+
+- [Operation Rename](#operation-rename "Operation Rename")
+- [Parameter Rename](#parameter-rename "Parameter Rename")
+- [Model Rename](#model-rename "Model Rename")
+- [Property Rename](#property-rename "Property Rename")
+- [Enum Value Rename](#enum-value-rename "Enum Value Rename")
+- [Change Description](#change-description "Change Description")
+
+### Operation Rename
+
+A common use case is renaming an operation. Say you have the following operation in your swagger:
+
+```yaml
+...
+"paths": {
+    "/vm": {
+        "get": {
+            "operationId": "getVirtualMachine",
+            ...
+        }
+    }
+    ...
+}
+```
+
+and you'd like to rename it from `getVirtualMachine` to `getVM`. You would refer to this operation with the combination of its `path` and its http `verb`.
+As always, you use `from: swagger-document` to specify you're changing your swagger document. Finally, what you're transforming is the `"operationId"`.
+Putting this together, your directive looks like:
+
+``` yaml
+### Directive renaming operation from "getVirtualMachine" to "getVM"
+directive:
+    from: swagger-document
+    where: '$.paths["/vm"].get'
+    transform: >
+        $["operationId"] = "getVM";
```

Comment:
SHould we use x-ms-clientName instead?

Reply:
that's fair, i'll change it to `x-ms-clientname` so it'll still work if opeartionis / params etc have an `x-ms-clientname` in their original definition